Sprint 11

Kirillov1515 opened the pull request.

Kirillov1515 committed:
Add ProgressHUD and UIBlockingProgressHUD

Kirillov1515 committed:
Introduce edits after the review

Kirillov1515 committed:
Реализовано задание 11 спринта, но показывается черный экран, вместо AuthViewController

Kirillov1515 committed:
Рабочая версия 11 спринта

Kirillov1515 committed:
Delete comments

Helvetii951 commented on ImageFeed/AuthAndWeb/AuthViewController.swift:
```diff
@@ -20,7 +20,9 @@ final class AuthViewController: UIViewController {
-            guard let webViewViewController = segue.destination as? WebViewViewController else { fatalError("Failed to prepare for \(showWebViewSegueIdentifire)") }
+            guard let webViewViewController = segue.destination as? WebViewViewController else { assertionFailure("Failed to prepare for \(showWebViewSegueIdentifire)")
+                return
+            }
```
Перед `assertionFailure` желательно сделать перенос строки.

Helvetii951 commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,74 @@
+        static let paramNameURL = "URL"
+    }
+    
+    //MARK: - Initialization
+    private init() {}
+    
```
За [singleton](https://theswiftdev.com/swift-singleton-design-pattern/) 👍

Helvetii951 commented on ImageFeed/Profile/ProfileImageService.swift:
```diff
@@ -0,0 +1,74 @@
+                    guard let avatarURL = avatarURL else { return }
+                    self.avatarURL = avatarURL
+                    completion(.success(avatarURL))
+                    NotificationCenter.default.post(name: ProfileImageService.didChangeNotification,
+                                                    object: self,
+                                                    userInfo: [Keys.paramNameURL: avatarURL])
```
В `userInfo` можно передать nil. Это значение дальше не используется.

Helvetii951 commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -48,29 +50,75 @@ final class ProfileViewController: UIViewController {
+        
+        let avatarPlaceholderImage = UIImage(named: "avatar_placeholder")
+        
+        let processor = RoundCornerImageProcessor(cornerRadius: 61)
+        avatarImageView.kf.indicatorType = .activity
+        avatarImageView.kf.setImage(
```
`RoundCornerImageProcessor` из `Kingfisher` скругляет именно ту картинку, которую скачал. Т.е., если он скачал большую картинку(например, 2000х2000), то он скруглит у нее только 61 поинт(такое число указано в коде проекта), что вряд ли будет заметно. `RoundCornerImageProcessor` никак не влияет на скругление UIImageView аватарки. Поэтому нужно скруглять `UIImageView` в коде контроллера профиля.

Helvetii951 commented on ImageFeed/ServiceAccessLayer/OAuth2/OAuth2Service.swift:
```diff
@@ -7,63 +7,65 @@
+    
     private (set) var authToken: String? {
         get {
-            return OAuth2TokenStorage().token
+            let tokenStorage = OAuth2TokenStorage()
+            return tokenStorage.token
         }
         set {
-            OAuth2TokenStorage().token = newValue
+            let tokenStorage = OAuth2TokenStorage()
+            tokenStorage.token = newValue
         }
     }
```
Рекомендую создать локальное свойство типа `OAuth2TokenStorage` и пользоваться им. Использование готового экземпляра по ресурсам дешевле, чем создание каждый раз нового.

Kirillov1515 replied:
Спасибо большое за все комментарии! Учту.

Helvetii951 commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -6,19 +6,41 @@
+        static let authViewControllerID = "AuthViewController"
+    }
+    
+    private let splashImageView: UIImageView = {
+        let imageView = UIImageView(image: UIImage(named: "launchScreen"))
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
+    }()
```
Рекомендую всегда использовать [ленивые свойства ](https://www.avanderlee.com/swift/lazy-var-property/)для объявления подобных свойств.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.

Kirillov1515 replied:
То есть имеет смысл всегда все вьюхи делать lazy?

Helvetii951 replied:
Это хорошая привычка. Иногда экономит ресурсы на создание и хранение, если, например, вьюха никогда не покажется в иерархии UI.

Но обрати внимание, что lazy всегда var. Т.е. свойство можно заменить на другой экземпляр. Так что ориентируйся по ситуации.

Kirillov1515 committed:
1. Исправил пул-реквест, теперь мерджится в main.
2. Проверил второе замечание - вроде работает корректно, сразу показывается реальная аватарка.
